packages/connector-tencent-email/src/index.ts

```diff
@@ -0,0 +1,150 @@
+import type {
+  CreateConnector,
+  EmailConnector,
+  GetConnectorConfig,
+  SendMessageFunction,
+} from '@logto/connector-kit';
+import {
+  ConnectorError,
+  ConnectorErrorCodes,
+  ConnectorType,
+  validateConfig,
+} from '@logto/connector-kit';
+import { assert } from '@silverhand/essentials';
+import { HTTPError } from 'got';
+
+import { isErrorResponse, request } from '@/http';
+import type { MailTemplateType, SendTencentMailConfig, ValueNames } from '@/types';
+import { sendTencentMailConfigGuard } from '@/types';
+
+import { defaultMetadata } from './constant';
+
+function getTemplate(templates: MailTemplateType[], usageType: string): MailTemplateType {
+  const template = templates.find((template) => template.usageType === usageType);
+  assert(
+    template,
+    new ConnectorError(
+      ConnectorErrorCodes.TemplateNotFound,
+      `Template not found for type: ${usageType}`
+    )
+  );
+
+  return template;
+}
+
+function buildFromName(fromName: string, fromAddress: string): string {
+  return fromName.length > 0 ? `${fromName} <${fromAddress}>` : fromAddress;
+}
+
+function buildTemplatePayload(
+  template: MailTemplateType,
+  templateParameters: Record<typeof ValueNames[number], string>
+): Record<string, string> {
+  return template.params.reduce<Record<string, string>>((accumulator, parameter) => {
+    const { name, value } = parameter;
+
+    return {
+      ...accumulator,
+      [name]: templateParameters[value],
+    };
+  }, {});
+}
+
+function transformString(...values: Array<string | undefined>): string {
+  return values.find((item) => item && item.length > 0) ?? '';
+}
+
+const sendMessage =
+  (getConfig: GetConnectorConfig): SendMessageFunction =>
+  async (data, inputConfig) => {
+    const { to, type, payload } = data;
+    const config = inputConfig ?? (await getConfig(defaultMetadata.id));
+    validateConfig<SendTencentMailConfig>(config, sendTencentMailConfigGuard);
+    const { accessKeyId, accessKeySecret, region, fromAddress, fromName, replyAddress, templates } =
+      config;
+
+    const mailTemplate: MailTemplateType = getTemplate(templates, type);
+
+    const fromAddressValue = transformString(mailTemplate.from?.fromAddress, fromAddress);
+
+    assert(
+      fromAddressValue.length > 0,
+      new ConnectorError(
+        ConnectorErrorCodes.InvalidConfig,
+        `mail params not found, fromAddress is required`
+      )
+    );
+
+    const fromNameValue = transformString(mailTemplate.from?.fromName, fromName);
+
+    const templateParameters: Record<typeof ValueNames[number], string> = {
+      code: payload.code,
+      toAddress: to,
+      fromAddress: fromAddressValue,
+      fromName: fromNameValue,
+      replayAddress: transformString(mailTemplate.from?.replyAddress, replyAddress),
+      subject: mailTemplate.subject ?? type,
```

Comment:
Can we just leave `subject` to empty if it is not provided? Should be an optional config, right?

Reply:
Yes, empty is already allowed in the default detection, but Tencent Cloud must provide the subject so here but when the user does not set it, the email type of logto is used as the subject.

Reply:
IMO, the email type might not be a proper value for the end users, as its value might be `SignIn` or `ForgotPassword`, right? How about just let it fallback to an empty string or a i18n placeholder like `No Subject`.

Reply:
Or if Tencent cloud requires subject, you can also make this property a required field in `MailTemplate` type, with zod guard to ensure the user must provide this in his config.